### ISSUE TITLE
Replace app-measurement.com in connectivity checks

### ISSUE THREAD

**Issue by bobobo1618** [firebase/firebase-ios-sdk#2254]
Continuation of https://github.com/invertase/react-native-firebase/issues/1469.

The problem: The SDK uses `app-measurement.com` to determine whether an internet connection is available. Apps utilizing this SDK sometimes operate in environments where users deliberately obstruct access to this domain through DNS filtering or IP-blocking, usually motivated by privacy concerns.

For example, [Pi-Hole](https://pi-hole.net/), a popular DNS server for privacy-sensitive users, has [a few default block lists](https://discourse.pi-hole.net/t/how-can-i-restore-pi-holes-default-ad-lists/4683/3?u=ramset), one of which is [StevenBlack/hosts](https://github.com/StevenBlack/hosts), which is an aggregation of these two lists which include `app-measurement.com` (per [the readme](https://github.com/StevenBlack/hosts/blob/master/readme.md)):
- [Steven Black's ad-hoc list](https://raw.githubusercontent.com/StevenBlack/hosts/master/data/StevenBlack/hosts)
- [Dan Pollock – someonewhocares](https://someonewhocares.org/hosts/zero/hosts)

Per the bug mentioned earlier, when the SDK attempts to connect to `app-measurement.com` and finds itself unable to do so, it appears to give up on making any other connections (like `<my-app>.firebase.com`), even when they would work. This can lead to an app becoming non-functional when it really shouldn't be.

As for solutions, some proposals:

- Treat this as a problem in the user's configuration: https://github.com/invertase/react-native-firebase/issues/1469#issuecomment-419725747
- Change the connectivity check hostname to `connectivitycheck.gstatic.com` or `captive.apple.com`: https://github.com/invertase/react-native-firebase/issues/1469#issuecomment-453729712 (my suggestion)

I'm not a fan of the first suggestion because:
- I believe we should try to respect users' wishes as much as possible. If they've chosen to opt-out of information being sent to this domain and there isn't a technical reason why access to the domain is needed, the app should continue to work.
- As users become more privacy-sensitive, environments like this are going to become more commonplace and it's unlikely analytics domains are going to be removed from blocklists (they're often an explicit goal of blocklists). There's a good chance a user wouldn't connect the use of privacy software with apps being broken, particularly since non-Firebase apps continue to work without issue. This will lead to confusion and a bad experience for users.

I propose the second suggestion because:
- It's dead simple. Just a couple of lines of code change.
- These domains are intended for connectivity checks and are built into mobile operating systems. If they don't work, the device shows error messaging to the user indicating their device is unable to connect. This makes it clear to the user that a particular network or setting is responsible for broken functionality. 
- The domains are single-purpose. All they do is determine whether the internet is accessible.
- Nothing is communicated to these domains as part of the connectivity check so privacy-sensitive users should have little reason to consider blocking them.
- These domains are baked into many wireless access points and as such are unlikely to change or become inaccessible in the near future.

Of the two domains:
- I suggested `connectivitycheck.gstatic.com` as it's controlled by Google, of which Firebase is a part and having control over the domain may give more confidence that it can be maintained in the future.
- I suggested `captive.apple.com` as this is an iOS SDK and privacy sensitive users may find it more acceptable that an app connect to Apple, as iOS will connect to Apple anyway as part of its own connectivity check.

As a user, I'd prefer the Apple domain.

**Comment by paulb777** [comment]
@bobobo1618 Thanks for the thorough Issue write-up. 

We've done a deeper investigation today and found that I don't yet fully understand the issue.  

FirebaseAnalytics should be the only Firebase library that uses GoogleUtilities/Network and `app-measurement.com` to check reachability.

FirebaseMessaging [checks for access to `www.google.com`](https://github.com/firebase/firebase-ios-sdk/blob/943141e0cd727211764ac871dead7c14ce38a5e7/Firebase/Messaging/FIRMessaging.m#L54).

FirebaseDatabase checks for [generic Internet reachability](https://github.com/firebase/firebase-ios-sdk/blob/4b3c511e8b31923ae6960b10bb41b710746042c0/Firebase/Database/Core/FPersistentConnection.m#L476) using an empty address.

With this information in mind, are you able to further narrow the root cause?

**Comment by bobobo1618** [comment]
Interesting, you're right.

When I did a search for `GULReachabilityChecker`, I came up with [FIRMessagingClient.m](https://github.com/firebase/firebase-ios-sdk/blob/98b6eef71eac368692ae249f056a75d89ed0350d/Firebase/Messaging/FIRMessagingClient.m) and [FIRMessaging.m](https://github.com/firebase/firebase-ios-sdk/blob/943141e0cd727211764ac871dead7c14ce38a5e7/Firebase/Messaging/FIRMessaging.m).

FIRMessaging.m, as you said, [initializes the reachability checker](https://github.com/firebase/firebase-ios-sdk/blob/943141e0cd727211764ac871dead7c14ce38a5e7/Firebase/Messaging/FIRMessaging.m#L278) with `www.google.com` as you say and FIRMessagingClient gets its reachability checker from FIRMessaging.

So it seems the failure in the bug may have a different root cause. Is it possible that the code making the analytics request just isn't failing gracefully? Is there code like:

```
doAnalytics();
doActualStuff();
```

That's failing to hit `doActualStuff` because `doAnalytics` failed?

It seems updating the connectivity check host may not fix the problem anyway so perhaps this should be a different bug.
